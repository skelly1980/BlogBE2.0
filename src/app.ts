import express, { NextFunction, Request, Response } from 'express';
import path from 'path';
import cookieParser from 'cookie-parser';
import cors from 'cors';
import helmet from 'helmet';
import morgan from 'morgan';
import apiRouter from './routes/apiRouter';
import viewRouter from './routes/viewRouter';
import { ApiError } from './error/ApiError';

const BASE_API_PATH = '/api';

export const app = express();

const logger = morgan('combined', {
  skip: () => process.env.NODE_ENV === 'test',
});

app.set('views', path.join(__dirname, 'views'));
app.set('view engine', 'ejs');

app.use(helmet());
app.use(cors());
app.use(logger);
app.use(express.json());
app.use(express.urlencoded({ extended: false }));
app.use(cookieParser());
app.use(express.static(path.join(__dirname, 'public')));

app.use(BASE_API_PATH, apiRouter);
app.use('/', viewRouter);

const isApiRequest = (req: Request) => {
  return req.url.startsWith(`${BASE_API_PATH}/`);
};

// 404 handler
app.use((req: Request, res: Response) => {
  res.status(404);

  if (isApiRequest(req)) {
    res.json({ message: 'Not found' });
    return;
  }

  res.render('404');
});

// api error handler
app.use((err: Error, req: Request, res: Response, next: NextFunction) => {
  if (err instanceof ApiError) {
    console.error(err.statusCode, err.message);
    res.status(err.statusCode).json({ message: err.message });
    return;
  }
  if (isApiRequest(req)) {
    console.error(err);
    res.status(500).json({ message: err.message });
    return;
  }
  next(err);
});

// view error handler
app.use((err: Error, _req: Request, res: Response, _next: NextFunction) => {
  res.status(500).render('500', { error: err.message });
});
