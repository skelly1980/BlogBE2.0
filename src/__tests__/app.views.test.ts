import fs from 'fs/promises';
import request from 'supertest';
import { app } from '../app';

describe('App views', () => {
  beforeEach(() => {
    jest.restoreAllMocks();
  });

  test('404', async () => {
    const response = await request(app).get('/invalid');
    expect(response.statusCode).toBe(404);
    expect(response.text).toMatchSnapshot();
  });

  describe('home page', () => {
    test('view', async () => {
      const response = await request(app).get('/');
      expect(response.statusCode).toBe(200);
      expect(response.text).toMatchSnapshot();
    });
  });

  describe('users page', () => {
    test('view', async () => {
      const response = await request(app).get('/users');
      expect(response.statusCode).toBe(200);
      expect(response.text).toMatchSnapshot();
    });

    test('throws error', async () => {
      jest.spyOn(fs, 'readFile').mockImplementation(() => {
        throw new Error('Kaboom!');
      });
      const response = await request(app).get('/users');
      expect(response.statusCode).toBe(500);
      expect(response.text).toMatchSnapshot();
    });
  });
});
