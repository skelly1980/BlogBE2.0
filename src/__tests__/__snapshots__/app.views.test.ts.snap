// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App views 404 1`] = `
"<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta content='width=device-width' name='viewport' />
  <title>Express with TypeScript Starter</title>
  <link rel="stylesheet" type="text/css" href="/stylesheets/styles.css" />
</head>

<body>
  Not Found
</body>

</html>"
`;

exports[`App views home page view 1`] = `
"<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta content='width=device-width' name='viewport' />
  <title>Express with TypeScript Starter</title>
  <link rel="stylesheet" type="text/css" href="/stylesheets/styles.css" />
</head>

<body>
  Hello World!
</body>

</html>"
`;

exports[`App views users page throws error 1`] = `
"<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta content='width=device-width' name='viewport' />
  <title>Express with TypeScript Starter</title>
  <link rel="stylesheet" type="text/css" href="/stylesheets/styles.css" />
</head>

<body>
  The following error occured: Kaboom!
</body>

</html>"
`;

exports[`App views users page view 1`] = `
"<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta content='width=device-width' name='viewport' />
  <title>Express with TypeScript Starter</title>
  <link rel="stylesheet" type="text/css" href="/stylesheets/styles.css" />
</head>

<body>
  <h2>Users</h2>
  <table>
    <thead>
      <tr>
        <th>ID</th>
        <th>Name</th>
        <th>Email</th>
      </tr>
    </thead>
    <tbody>
      
      <tr>
        <td>1</td>
        <td>Bob Example</td>
        <td>bob@example.com</td>
      </tr>
      
      <tr>
        <td>2</td>
        <td>Daffy Duck</td>
        <td>d.duck@wb.com</td>
      </tr>
      
      <tr>
        <td>3</td>
        <td>Geddy Lee</td>
        <td>dirk@rush.com</td>
      </tr>
      
      <tr>
        <td>4</td>
        <td>Paul McCartney</td>
        <td>p.mccartney@beatles.com</td>
      </tr>
      
    </tbody>
  </table>
</body>

</html>"
`;
